src/helics/core/zmq/ZmqCommsSS.cpp

```diff
@@ -0,0 +1,543 @@
+                        {
+                        	brokerConnection.send (buffer.data (), buffer.size (), ZMQ_NOBLOCK);
+                        }
+                        else
+                        {
+                            logWarning ("unknown route and no broker, dropping message");
+                        }
+                    }
+                }
+            }
+            tx_count++;
+            if (tx_count < TX_RX_MSG_COUNT) {
+            	tx_msg = txQueue.try_pop ();
+            }
+        }
+
+        count = 0;
+        rc = 1;
+        while ((rc > 0) && (count < 5))
+        {
+        	rc = zmq::poll (poller, std::chrono::milliseconds (10));
+
+            if (rc > 0)
+            {
+                if ((poller[0].revents & ZMQ_POLLIN) != 0)
+                {
+                    status = processRxMessage (brokerSocket, brokerConnection, connection_info);
+                }
+                if (serverMode && hasBroker)
+                {
+                    if ((poller[1].revents & ZMQ_POLLIN) != 0)
+                    {
+                        status = processRxMessage (brokerSocket, brokerConnection, connection_info);
+                    }
+                }
+
+                if (status < 0)
+                {
+                	goto CLOSE_TX_LOOP;
+                }
+            }
+            count++;
+        }
+    }
+    CLOSE_TX_LOOP:
+		routes.clear ();
+		connection_info.clear ();
+		if (getRxStatus () == connection_status::connected)
+		{
+			setRxStatus (connection_status::terminated);
+		}
+		if (serverMode)
+		{
+			std::this_thread::sleep_for (std::chrono::milliseconds (50));
+			brokerSocket.close ();
+		}
+		if (hasBroker)
+		{
+			brokerConnection.close ();
+		}
+		setTxStatus (connection_status::terminated);
+}
+
+int ZmqCommsSS::processRxMessage (zmq::socket_t &brokerSocket,
+                                    zmq::socket_t &brokerConnection,
+                                    std::map<std::string, std::string> &connection_info)
+{
+    int status = 0;
+    zmq::message_t msg1;
+    zmq::message_t msg2;
+
+    if (serverMode)
+    {
+        brokerSocket.recv (&msg1);
+        std::string str (static_cast<char *> (msg1.data ()), msg1.size ());
+
+        brokerSocket.recv (&msg2);
+    }
+    else
+    {
+        brokerConnection.recv (&msg1);
+        std::string str (static_cast<char *> (msg1.data ()), msg1.size ());
+
+        brokerConnection.recv (&msg2);
+    }
+    std::string str2 (static_cast<char *> (msg2.data ()), msg2.size ());
+    status = processIncomingMessage (msg2, connection_info);
+
+    return status;
+}
+
+void ZmqCommsSS::closeReceiver ()
+{
+    switch (getTxStatus ())
+    {
+    case connection_status::startup:
+    case connection_status::connected:
+    {
+    	ActionMessage cmd (CMD_PROTOCOL);
+        cmd.messageID = CLOSE_RECEIVER;
+        transmit (control_route, cmd);
+    }
+
+    break;
+    default:
+        if (!disconnecting)
```

Comment:
If the code gets here I would guess this is not going to work, since you are doing the same thing as if it were connected.  But in this case that probably doesn't matter since there is only one thread and the receiver and transmitter should have the same close

Reply:
Removed the check now